File: psana_examples/tags/V00-00-10/src/DumpPnccd.cpp
//--------------------------------------------------------------------------
// File and Version Information:
// 	$Id$
//
// Description:
//	Class DumpPnccd...
//
// Author List:
//      Andrei Salnikov
//
//------------------------------------------------------------------------

//-----------------------
// This Class's Header --
//-----------------------
#include "psana_examples/DumpPnccd.h"

//-----------------
// C/C++ Headers --
//-----------------

//-------------------------------
// Collaborating Class Headers --
//-------------------------------
#include "MsgLogger/MsgLogger.h"
#include "psddl_psana/pnccd.ddl.h"

//-----------------------------------------------------------------------
// Local Macros, Typedefs, Structures, Unions and Forward Declarations --
//-----------------------------------------------------------------------

using namespace psana_examples;
PSANA_MODULE_FACTORY(DumpPnccd)

//		----------------------------------------
// 		-- Public Function Member Definitions --
//		----------------------------------------

namespace psana_examples {

//----------------
// Constructors --
//----------------
DumpPnccd::DumpPnccd (const std::string& name)
  : Module(name)
{
    m_src = configStr("source", "DetInfo(:pnCCD)");
}

//--------------
// Destructor --
//--------------
DumpPnccd::~DumpPnccd ()
{
}

/// Method which is called at the beginning of the calibration cycle
void 
DumpPnccd::beginCalibCycle(Env& env)
{
  MsgLog(name(), trace, "in beginCalibCycle()");

  shared_ptr<Psana::PNCCD::ConfigV1> config1 = env.configStore().get(m_src);
  if (config1.get()) {
    
    WithMsgLog(name(), info, str) {
      str << "PNCCD::ConfigV1:";
      str << "\n  numLinks = " << config1->numLinks();
      str << "\n  payloadSizePerLink = " << config1->payloadSizePerLink();
    }
    
  }

  shared_ptr<Psana::PNCCD::ConfigV2> config2 = env.configStore().get(m_src);
  if (config2.get()) {
    
    WithMsgLog(name(), info, str) {
      str << "PNCCD::ConfigV2:";
      str << "\n  numLinks = " << config2->numLinks();
      str << "\n  payloadSizePerLink = " << config2->payloadSizePerLink();
      str << "\n  numChannels = " << config2->numChannels();
      str << "\n  numRows = " << config2->numRows();
      str << "\n  numSubmoduleChannels = " << config2->numSubmoduleChannels();
      str << "\n  numSubmoduleRows = " << config2->numSubmoduleRows();
      str << "\n  numSubmodules = " << config2->numSubmodules();
      str << "\n  camexMagic = " << config2->camexMagic();
      str << "\n  info = " << config2->info();
      str << "\n  timingFName = " << config2->timingFName();
    }
    
  }

}

/// Method which is called with event data, this is the only required 
/// method, all other methods are optional
void 
DumpPnccd::event(Event& evt, Env& env)
{
  shared_ptr<Psana::PNCCD::FrameV1> frame = evt.get(m_src);
  if (frame.get()) {
    WithMsgLog(name(), info, str) {
      str << "PNCCD::FrameV1:";
      str << "\n  specialWord = " << frame->specialWord();
      str << "\n  frameNumber = " << frame->frameNumber();
      str << "\n  timeStampHi = " << frame->timeStampHi();
      str << "\n  timeStampLo = " << frame->timeStampLo();

      const uint16_t* data = frame->data();
      str << "\n  data =";
      for (int i = 0; i < 10; ++ i) {
        str << " " << data[i];
      }
      str << " ...";
    }
  }
}

} // namespace psana_examples

File: python_examples/trunk/src/dump_pnccd.py
#--------------------------------------------------------------------------
# File and Version Information:
#  $Id: dump_pnccd.py 2622 2011-11-11 14:35:00Z salnikov@SLAC.STANFORD.EDU $
#
# Description:
#  Module myana...
#
#------------------------------------------------------------------------

"""User analysis job for XTC data.

This software was developed for the LUSI project.  If you use all or 
part of it, please give an appropriate acknowledgment.

@see RelatedModule

@version $Id: dump_pnccd.py 2622 2011-11-11 14:35:00Z salnikov@SLAC.STANFORD.EDU $

@author Andrei Salnikov
"""


#------------------------------
#  Module's version from CVS --
#------------------------------
__version__ = "$Revision: 2622 $"
# $Source$

#--------------------------------
#  Imports of standard modules --
#--------------------------------
import sys
import logging

#---------------------------------
#  Imports of base class module --
#---------------------------------

#-----------------------------
# Imports for other modules --
#-----------------------------
from pypdsdata import *

#----------------------------------
# Local non-exported definitions --
#----------------------------------

#------------------------
# Exported definitions --
#------------------------

#---------------------
#  Class definition --
#---------------------
class dump_pnccd ( object ) :
    """Example analysis module which dumps pnCCD info."""

    #--------------------
    #  Class variables --
    #--------------------

    #----------------
    #  Constructor --
    #----------------
    def __init__ ( self, source="" ) :
        """Class constructor takes the name of the data source.

        @param source   data source
        """
        
        self.m_src = source

    #-------------------
    #  Public methods --
    #-------------------

    def beginjob( self, evt, env ) :

        config = env.getConfig(xtc.TypeId.Type.Id_pnCCDconfig, self.m_src)
        if config:
        
            print "%s: %s" % (config.__class__.__name__, self.m_src)
            print "  numLinks =", config.numLinks()
            print "  payloadSizePerLink =", config.payloadSizePerLink()
            
            try:
                #these methods exist only in V2
                print "  numChannels = %s" % config.numChannels()
                print "  numRows =", config.numRows()
                print "  numSubmoduleChannels =", config.numSubmoduleChannels()
                print "  numSubmoduleRows =", config.numSubmoduleRows()
                print "  numSubmodules =", config.numSubmodules()
                print "  camexMagic =", config.camexMagic()
                print "  info =", config.info()
                print "  timingFName =", config.timingFName()
            except:
                pass

    def beginrun( self, evt, env ) :
        pass

    def event( self, evt, env ) :
        
        frame = evt.get(xtc.TypeId.Type.Id_pnCCDframe, self.m_src)
        if frame:
            print "%s: %s" % (frame.__class__.__name__, self.m_src)
            print "  specialWord =", frame.specialWord()
            print "  frameNumber =", frame.frameNumber()
            print "  timeStampHi =", frame.timeStampHi()
            print "  timeStampLo =", frame.timeStampLo()
            data = frame.data()
            print "  data.shape =", data.shape
            print "  data.dtype =", data.dtype
            print "  data =", data
        
        
    def endrun( self, env ) :
        pass

    def endjob( self, env ) :
        pass
